Configure TCP or UNIX socket

mamedin opened the pull request:
A new variable `CLAMAV_SOCKET_TYPE` has been added. It can take two
values: "tcp" or "file".

On CentOS and Ubuntu 14 it is configured in the `clamd.conf` file.

On Ubuntu 16 this configuration is more complex because the
clamav-daemon package uses three systemd files to configure the socket
type (unix file or tcp socket). This PR uses debconf-set-selections to
configure the socket type, but the clamav-daemon package doesn't read
the debconf system when upgrading or reinstalling, it is used only when
installing from scratch, so a package purge is needed when changing the
socket type.

The SELinux policies `antivirus_can_scan_system` and `clamd_use_jit` will
be activated if SELinux is enabled on CentOS.

On Ubuntu task files, the "Enable services" task has been modified to
restart the clamav-daemon service instead of starting. It is needed to
apply the changes made by the role.

On Ubuntu the default socket type is UNIX file.
On Centos the default socket type is TCP.

Finally, 2 additional variables have been added to define the TCP port and IP address to listen on:

- CLAMAV_TCP_ADDR: Default 127.0.0.1
- CLAMAV_TCP_PORT: Default 3310

It Fixes #5

mamedin committed:
Configure TCP or UNIX socket

A new variable `CLAMAV_SOCKET_TYPE` has been added. It can take two
values: "tcp" or "file".

On CentOS and Ubuntu 14 it is configured in the `clamd.conf` file.

On Ubuntu 16 this configuration is more complex because the
clamav-daemon package uses three systemd files to configure the socket
type (unix file or tcp socket). This PR uses debconf-set-selections to
configure the socket type, but the clamav-daemon package doesn't read
the debconf system when upgrading or reinstalling, it is used only when
installing from scratch, so a package purge is needed when changing the
socket type.

The SELinux policies `antivirus_can_scan_system` and `clamd_use_jit` will
be activated if SELinux is enabled on CentOS.

On Ubuntu task files, the "Enable services" task has been modified to
restart the clamav-daemon service instead of starting. It is needed to
apply the changes made by the role.

On Ubuntu the default socket type is UNIX file.
On Centos the default socket type is TCP.

scollazo commented on README.md:
```diff
@@ -3,8 +3,15 @@ ansible-clamav
+______________
+
+- `CLAMAV_SOCKET_TYPE`: configure clamad to use a tcp or file socket (default: "file")
+- `CLAMAV_MAX_FILE_SIZE`: (default: "2000M")
```
Those fields should also have an explanation, not only `CLAMAV_SOCKET_TYPE`

scollazo commented on README.md:
```diff
@@ -3,8 +3,15 @@ ansible-clamav
+Role Variables
+______________
+
+- `CLAMAV_SOCKET_TYPE`: configure clamad to use a tcp or file socket (default: "file")
```
It reads `clamad`, typo?

scollazo commented on defaults/main.yml:
```diff
@@ -5,3 +5,9 @@ CLAMAV_MAX_FILE_SIZE: "2000M"
 CLAMAV_MAX_STREAM_LENGTH: "2000M"
 
+# The clamvd can be configured to work with a file socket or a tcp socket.
+# Available Options for CLAMAV_SOCKET_TYPE are "tcp" and "file".
```
`clamvd`?

scollazo commented on defaults/main.yml:
```diff
@@ -5,3 +5,9 @@ CLAMAV_MAX_FILE_SIZE: "2000M"
+# The clamvd can be configured to work with a file socket or a tcp socket.
+# Available Options for CLAMAV_SOCKET_TYPE are "tcp" and "file".
+# This env var has not a default value. 
+# On CentOS the role will use "tcp" if not defined
```
Itemize this?

```
#  - On CentOS...
#  - On Ubuntu...
```

scollazo commented on tasks/main.yml:
```diff
@@ -5,7 +5,13 @@
   static: "no"
   when:
-    - ansible_distribution == "Ubuntu"
+    - ansible_distribution == "Ubuntu" and ansible_distribution_release != "trusty"
```
May be we should be explicit here, and use when `ansible_distribution_release= "xenial"`. Did you test this with newer Ubuntu versions (18.04 is almost out!)

scollazo commented on tasks/redhat.yml:
```diff
@@ -54,9 +73,17 @@
+  shell: "systemd-tmpfiles --create /etc/tmpfiles.d/clamd.scam.archivematica.conf"
+  when: CLAMAV_SOCKET_TYPE ==  'file'
+
+- name: "Enable and restart services"
```
This should be triggered only when a configuration change is made

scollazo commented on tasks/ubuntu.yml:
```diff
@@ -7,6 +7,66 @@
+# clamav-daemon already installed block
+- block:
+
+   # check if clamav is running with file or tcp socket
```
First letter should be capital

scollazo commented on tasks/ubuntu.yml:
```diff
@@ -7,6 +7,66 @@
+       clamav_uses_file_socket: true
+     when: clamavunixsocket is defined and ( clamavunixsocket.rc == 0 )
+
+   # purge clamav-daemon if was installed and socket type will change
```
First letter should be capital

scollazo commented on tasks/ubuntu.yml:
```diff
@@ -7,6 +7,66 @@
+       - CLAMAV_SOCKET_TYPE !=  'tcp' and clamav_uses_file_socket is not defined or clamav_daemon_already_installed is defined and CLAMAV_SOCKET_TYPE == 'tcp' and clamav_uses_file_socket is defined
+  when: clamav_daemon_already_installed is defined
+
+# set debconf when installing clamav-daemon
```
...you know what I'm about to say here...

scollazo commented on tasks/ubuntu.yml:
```diff
@@ -39,7 +99,7 @@
 - name: "Enable services"
```
As with Centos, this should be triggered only on configuration change

scollazo commented on templates/etc/clamd.d/scan.conf.j2:
```diff
@@ -99,15 +104,23 @@ LogSyslog yes
 # times if you want to listen on multiple IPs. IPv6 is now supported.
 # Default: no
+{% if CLAMAV_SOCKET_TYPE == 'file' %}
+#TCPAddr 127.0.0.1
```
Can the tcp bind address and port be converted to vars? There are environments where we need to have clamav listening on a different port, due to 3310 being used by other service.

mamedin commented:
@scollazo  thanks for your suggestions. I have changed all that you said except triggering the clamav-daemon restart only on configuration changes. I think I have added a lot of complexity to this role, especially in ubuntu xenial and the clamd is a very light demon that can be restarted without any problem.

hakamine committed:
Add travis config

Check ansible syntax and run playbook on container

hakamine reviewed (approved):
:+1: I added a commit with travis config to run a syntax check on the role, and do a test run with the default configuration, on the currently 3 supported OSes, tests are passing.

sevein commented:
> 👍 I added a commit with travis config to run a syntax check on the role, and do a test run with the default configuration, on the currently 3 supported OSes, tests are passing.

🎉 this is interesting, @hakamine. So there's not test assertions other than expecting that the playbook completes successfully, right? I guess that's a good start.
